library/ecp.c

```diff
@@ -1958,6 +1958,64 @@ int mbedtls_ecp_check_pub_priv( const mbedtls_ecp_keypair *pub, const mbedtls_ec
     return( ret );
 }
 
+/*
+ *  Decompresses an EC Public Key
+ */
+int mbedtls_ecp_decompress_pubkey( const mbedtls_ecp_group *grp, const unsigned char *input, size_t ilen, unsigned char *output, size_t *olen, size_t osize ){
+
+  int ret;
+  mbedtls_mpi x, x3, ax, z, zexp, y;
+  size_t plen;
+
+  plen = mbedtls_mpi_size( &grp->P );
+  *olen = 2 * plen + 1;
+
+  if( ilen != plen + 1 || input[0] == 0x04 )
+    return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
+
+  if( osize < 1 )
+    return( 0 );
+
+  if( osize < *olen )
+    return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+
+  mbedtls_mpi_init( &x );
+  mbedtls_mpi_init( &x3 );
+  mbedtls_mpi_init( &ax );
+  mbedtls_mpi_init( &z );
+  mbedtls_mpi_init( &zexp );
+  mbedtls_mpi_init( &y );
+
+
+  MBEDTLS_MPI_CHK( mbedtls_mpi_read_binary( &x, input+1, ilen-1 ) ); // X point of the pubkey
+  MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &x3, &x, &x ) ); // X^2
+  MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &x3, &x3, &x ) ); // X^3
+  MBEDTLS_MPI_CHK( mbedtls_mpi_mul_mpi( &ax, &grp->A, &x ) ); // AX
+  MBEDTLS_MPI_CHK( mbedtls_mpi_add_mpi( &z, &x3, &ax ) ); // X^3 + AX
+  MBEDTLS_MPI_CHK( mbedtls_mpi_add_mpi( &z, &z, &grp->B ) ); // X^3 + AX + B
+  MBEDTLS_MPI_CHK( mbedtls_mpi_copy( &zexp, &grp->P) ); // Z exponent
+  MBEDTLS_MPI_CHK( mbedtls_mpi_add_int( &zexp, &zexp, 1 ) ); // Z exponent + 1
+  MBEDTLS_MPI_CHK( mbedtls_mpi_div_int( &zexp, 0, &zexp, 4 ) ); // Z exponent / 4
+  MBEDTLS_MPI_CHK( mbedtls_mpi_exp_mod( &y, &z, &zexp, &grp->P, 0 ) ); // Z^Zexp % P
```

Comment:
This way of extracting square roots mod P only works for some Ps (those that are congruent to 3 mod 4), but will not work for all curves supported by Mbed TLS.

Reply:
@mpg what curves use Ps that are congruent to 3 mod 4? I would like to add that to my readme here: https://github.com/mwarning/mbedtls_ecp_compression

Reply:
It's easy to check:
```c
#include "mbedtls/ecp.h"
#include <stdio.h>

int main(void)
{
    const mbedtls_ecp_curve_info *info;
    mbedtls_ecp_group grp;
    mbedtls_mpi_uint r;

    for( info = mbedtls_ecp_curve_list(); info->name != NULL; info++ )
    {
        mbedtls_ecp_group_init( &grp );
        mbedtls_ecp_group_load( &grp, info->grp_id );
        mbedtls_mpi_mod_int( &r, &grp.P, 4 );
        printf( "%s: p = %u mod 4\n", info->name, (unsigned) r );
        mbedtls_ecp_group_free( &grp );
    }
}
```
then
```
% make lib && gcc -Iinclude p-mod-4.c library/libmbedcrypto.a -o p-mod-4 && ./p-mod-4                                                                  git|development|
secp521r1: p = 3 mod 4
brainpoolP512r1: p = 3 mod 4
secp384r1: p = 3 mod 4
brainpoolP384r1: p = 3 mod 4
secp256r1: p = 3 mod 4
secp256k1: p = 3 mod 4
brainpoolP256r1: p = 3 mod 4
secp224r1: p = 1 mod 4
secp224k1: p = 1 mod 4
secp192r1: p = 3 mod 4
secp192k1: p = 3 mod 4
```